namespace.go

```diff
@@ -231,12 +238,56 @@ func (ns *Namespace) newClient(ctx context.Context) (*amqp.Client, error) {
 	return amqp.Dial(ns.getAMQPHostURI(), defaultConnOptions...)
 }
 
-func (ns *Namespace) negotiateClaim(ctx context.Context, client *amqp.Client, entityPath string) error {
+// negotiateClaim performs initial authentication and starts periodic refresh of credentials.
+// the returned func is to cancel() the refresh goroutine.
+func (ns *Namespace) negotiateClaim(ctx context.Context, client *amqp.Client, entityPath string) (func() <-chan struct{}, error) {
 	ctx, span := ns.startSpanFromContext(ctx, "sb.namespace.negotiateClaim")
 	defer span.End()
 
 	audience := ns.getEntityAudience(entityPath)
-	return cbs.NegotiateClaim(ctx, audience, client, ns.TokenProvider)
+	if err := cbs.NegotiateClaim(ctx, audience, client, ns.TokenProvider); err != nil {
+		return nil, err
+	}
+	ns.initRefresh.Do(func() {
+		// start the periodic refresh of credentials
+		refreshCtx, done := context.WithCancel(context.Background())
+		exitChan := make(chan struct{})
+		go func() {
+			defer func() {
+				// reset the guard when the refresh goroutine exits
+				ns.initRefresh = sync.Once{}
+				// signal that the refresh goroutine has exited
+				close(exitChan)
+			}()
+			const refreshDelay = 15 * time.Minute
+			timer := time.NewTimer(refreshDelay)
+			for {
+				select {
+				case <-refreshCtx.Done():
+					// stop timer and drain channel
+					if !timer.Stop() {
+						<-timer.C
+					}
+					return
+				case <-timer.C:
```

Comment:
what about using 

```golang
<- timeAfter(refreshDelay)
```

instead of instanciated a timer and resetting it?

Reply:
I agree with this change. It would reduce the code complexity a lot, and it will be less error-prone.

Reply:
That's what I originally had, however I changed it to what's there now after reading the docs for `time.After()`.

```go
// The underlying Timer is not recovered by the garbage collector
// until the timer fires.
...
```

Given that we regularly call `Recover()` to rebuild the connections (which will likely exit the refresh goroutine before the timer fires) I figured using the timer directly would let us control the clean-up.  I admit this is relatively a minor issue so if folks still think the extra complexity isn't worth it I can change it.

Reply:
```golang
// until the timer fires. If efficiency is a concern, use NewTimer
// instead and call Timer.Stop if the timer is no longer needed.
...
```

I don't think this kind of optimization is necessary, if it was a concern for an application, then probably it wouldn't use this library or even Service Bus. This kind of optimization is nice for embedded systems and other systems with extremely low resources, not for the usual (web) applications implementing this library. For the sake of simplicity, I would change the code to use `time.After` instead.

Reply:
Switched to using `time.After()`